fuzz/ml-dsa.c

```diff
@@ -0,0 +1,624 @@
+ *       and should be freed appropriately using `EVP_PKEY_free()`.
+ */
+static void create_ml_dsa_raw_key(uint8_t **buf, size_t *len,
+                                  void **key1, void **key2)
+{
+    EVP_PKEY *pubkey;
+    char *keytype = NULL;
+    size_t keylen = 0;
+    /* MAX_ML_DSA_PRIV_LEN is longer of that and ML_DSA_87_PUB_LEN */
+    uint8_t key[MAX_ML_DSA_PRIV_LEN];
+    int pub = 0;
+
+    if (!select_keytype_and_size(buf, len, &keytype, &keylen, 0))
+        return;
+
+    /*
+     * Select public or private key creation based on the low order bit of the
+     * next buffer value.
+     * Note that keylen as returned from select_keytype_and_size is a public key
+     * length, so make the adjustment to private key lengths here.
+     */
+    if ((*buf)[0] & 0x1) {
+        pub = 1;
+    } else {
+        switch (keylen) {
+        case (ML_DSA_44_PUB_LEN):
+            keylen = ML_DSA_44_PRIV_LEN;
+            break;
+        case (ML_DSA_65_PUB_LEN):
+            keylen = ML_DSA_65_PRIV_LEN;
+            break;
+        case (ML_DSA_87_PUB_LEN):
+            keylen = ML_DSA_87_PRIV_LEN;
+            break;
+        default:
+            return;
+        }
+    }
+
+    /*
+     * libfuzzer provides by default up to 4096 bit input buffers, but it's
+     * typically much less (between 1 and 100 bytes) so use RAND_bytes here
+     * instead
+     */
+    if (!RAND_bytes(key, keylen))
+        return;
+
+    /*
+     * Try to generate either a raw public or private key using random data
+     * Because the input is completely random, it's effectively certain this
+     * operation will fail, but it will still exercise the code paths below,
+     * which is what we want the fuzzer to do
+     */
+    if (pub == 1)
+        pubkey = EVP_PKEY_new_raw_public_key_ex(NULL, keytype, NULL, key, keylen);
+    else
+        pubkey = EVP_PKEY_new_raw_private_key_ex(NULL, keytype, NULL, key, keylen);
+
+    *key1 = pubkey;
+    return;
+}
+
+/**
+ * @brief Generates a valid ML-DSA key using OpenSSL.
+ *
+ * This function selects a valid ML-DSA key type and size from the buffer,
+ * initializes an OpenSSL EVP_PKEY context, and generates a cryptographic key
+ * accordingly.
+ *
+ * @param buf    Pointer to the buffer pointer; updated after reading.
+ * @param len    Pointer to the remaining buffer size; updated accordingly.
+ * @param key1   Pointer to store the generated EVP_PKEY key.
+ * @param unused Unused parameter (reserved for future use).
+ *
+ * @note The generated key is allocated using OpenSSL's EVP_PKEY functions
+ *       and should be freed using `EVP_PKEY_free()`.
+ */
+static void keygen_ml_dsa_real_key(uint8_t **buf, size_t *len,
+                                   void **key1, void **key2)
+{
+    char *keytype = NULL;
+    size_t keylen = 0;
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY **key;
+
+    *key1 = *key2 = NULL;
+
+    key = (EVP_PKEY **)key1;
+
+again:
```

Comment:
Why so convoluted?  Why not just call a function that creates a single key twice, once with key1 and separately with key2?

Reply:
This was just copied exactly from https://github.com/openssl/openssl/pull/26657, but I'll update it to call a helper function twice

Reply:
Thats my bad, just happened to be how I wrote it, as for the ML-KEM test I typically needed two keys for most tests.  It should be a helper function